### PR TITLE
Replace proc-macro-error proc-macro2-diagnostics

### LINKED ISSUE #61: Replace proc-macro-error with proc-macro2-diagnostics
`proc-macro-error` which is used for pretty errors [seems to be unmaintained](https://gitlab.com/CreepySkeleton/proc-macro-error/-/issues/20) and also [still depends on Syn 1.x](https://gitlab.com/CreepySkeleton/proc-macro-error/-/merge_requests).

The latter is my main driver for opening this issue as I'm [working on cutting down on duplicate dependencies](https://github.com/janhohenheim/foxtrot/pull/311) in @janhohenheim's Bevy game template [Foxtrot](https://github.com/janhohenheim/foxtrot). His template is quite representative of the kinds of dependencies that you'll find in Bevy games.

I propose replacing it with [proc-macro2-diagnostics](https://github.com/SergioBenitez/proc-macro2-diagnostics). Doing so is quite easy (I have a PR ready).

### PR DESCRIPTION
**Pull request by bash**
#### Fixes #61

### CODE REVIEW

**Comment by Emoun** [comment]
@bash, would you be able to investigate what the nightly error is in CI?

**Comment by bash** [comment]
Sure thing :) 

From a quick glance it seems that the difference comes from the fact that on nightly `proc-macro`'s built-in diagnostics are used whereas on stable a shim is used instead. The two impls format *slightly* differently.

Would you be ok with separate "expected" files for stable vs. nightly?

**Comment by Emoun** [comment]
I would dislike that very much. What is the difference? Could we solve it using a less strict comparison in those tests?
and is there really no way to have it behave the same on both stable and nightly?

**Comment by bash** [comment]
That's fair enough :) I think it's some minor differences (indentation, prefix). I'll take a closer look though

File: Cargo.toml
Comment (Emoun) [inline Cargo.toml:33]: What does `dep:` do here?

File: Cargo.toml
Comment (Emoun) [inline Cargo.toml:36]: No need for backwards compatibility since we already have other breaking changes.
```suggestion
```

File: src/error.rs
Comment (Emoun) [inline src/error.rs:72]: Please add relevant doc comment

**Review by Emoun — CHANGES_REQUESTED** [review]
@bash, thanks for the PR.

I have added some comments that I hope you can address when you have time.

File: Cargo.toml
Comment (bash) [reply Cargo.toml:33]: `dep:` stops the optional dependency from becoming an implicit feature.

From the [Cargo Book](https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies):
> In some cases, you may not want to expose a feature that has the same name as the optional dependency. For example, perhaps the optional dependency is an internal detail, or you want to group multiple optional dependencies together, or you just want to use a better name. If you specify the optional dependency with the `dep:` prefix anywhere in the `[features]` table, that disables the implicit feature.

File: src/error.rs
Comment (bash) [reply src/error.rs:78]: d5520cf78c2674d05e56d5961f7a68163e996467

File: src/error.rs
Comment (bash) [reply src/error.rs]: d5520cf78c2674d05e56d5961f7a68163e996467

File: src/error.rs
Comment (bash) [reply src/error.rs]: d5520cf78c2674d05e56d5961f7a68163e996467

File: src/error.rs
Comment (bash) [reply src/error.rs]: d5520cf78c2674d05e56d5961f7a68163e996467

File: src/error.rs
Comment (bash) [reply src/error.rs]: d5520cf78c2674d05e56d5961f7a68163e996467

File: tests/errors/hint/duplicate_without_duplicates
Comment (bash) [reply tests/errors/hint/duplicate_without_duplicates:1]: Hmm I don't think so. The prefix comes from the `Diagnostic`'s level as far as I can tell.

**Comment by bash** [comment]
> That's fair enough :) I think it's some minor differences (indentation, prefix). I'll take a closer look though

The difference just boils down to indentation if we just look at the help.
Relaxing the duplicate_macrotest to ignore leading/trailing whitespace on each line makes the tests pass :)

**stable**
```
error: Wrong argument count for substitution identifier.
       = help: The same substitution identifier must take the same number of argument across all substitution groups.
       Example:
           [
               ident1(arg1, arg2)  [sub1 arg1 arg2]
                      ^^^^^^^^^^
           ]
           [
               ident1(arg1, arg2)  [arg1 arg2 sub2]
                      ^^^^^^^^^^
           ]
       
  --> tests/tests.rs:15:3
   |
15 |         name     [sub1]
   |         ^^^^

```

**nightly**
```
error: Wrong argument count for substitution identifier.
  --> tests/tests.rs:15:3
   |
15 |         name     [sub1]
   |         ^^^^
   |
   = help: The same substitution identifier must take the same number of argument across all substitution groups.
           Example:
               [
                   ident1(arg1, arg2)  [sub1 arg1 arg2]
                          ^^^^^^^^^^
               ]
               [
                   ident1(arg1, arg2)  [arg1 arg2 sub2]
                          ^^^^^^^^^^
               ]
```

File: Cargo.toml
Comment (Emoun) [reply Cargo.toml:33]: Wow I didn't know about this. Thank you for using it and getting it to my attention. I think this probably needs a mention in the changelog

**Review by Emoun — COMMENTED** [review]
Thanks for the changes. I think it's ready to merge, however, I'll wait for emoun/duplicate_macrotest#1 to be merged and released before doing a last manual check and then merging this.

**Comment by Emoun** [comment]
@bash thank you for all the work on this PR. Everything seems to be working now, so I will merge this.

I don't know exactly when I'm going to be able to release a version including these changes, but I'll do my best.

**Comment by bash** [comment]
> I don't know exactly when I'm going to be able to release a version including these changes, but I'll do my best.

Take your time 😊 
I'm just happy that syn 1.x will disappear from my dependency tree *eventually* :)

**Pull request merged**
